In [31]:
import numpy as np
import pandas as pd
import plotly.graph_objects as go
import plotly.io as pio
import plotly.express as px

#Load dữ liệu

In [32]:
from sklearn.datasets import load_iris
iris=load_iris()

In [33]:
iris

{'DESCR': '.. _iris_dataset:\n\nIris plants dataset\n--------------------\n\n**Data Set Characteristics:**\n\n    :Number of Instances: 150 (50 in each of three classes)\n    :Number of Attributes: 4 numeric, predictive attributes and the class\n    :Attribute Information:\n        - sepal length in cm\n        - sepal width in cm\n        - petal length in cm\n        - petal width in cm\n        - class:\n                - Iris-Setosa\n                - Iris-Versicolour\n                - Iris-Virginica\n                \n    :Summary Statistics:\n\n    ============== ==== ==== ======= ===== ====================\n                    Min  Max   Mean    SD   Class Correlation\n    ============== ==== ==== ======= ===== ====================\n    sepal length:   4.3  7.9   5.84   0.83    0.7826\n    sepal width:    2.0  4.4   3.05   0.43   -0.4194\n    petal length:   1.0  6.9   3.76   1.76    0.9490  (high!)\n    petal width:    0.1  2.5   1.20   0.76    0.9565  (high!)\n    ===========

Các label được encode như sau:

• 0: Setosa

• 1: Versicolour

• 2: Virginica

In [34]:
X=iris.data[:,:2]
y=iris.target

In [35]:
X.shape

(150, 2)

In [36]:
y.shape

(150,)

#Biểu đồ phân tích dữ liệu

In [37]:
pio.templates.default = "plotly_dark"
fig=px.pie(values=[(y==0).sum(), (y==1).sum(), (y==2).sum()], names=iris.target_names, title="Distribution of Classes")
fig.show()

Tập dataset có tỉ lệ giữa các class là cân bằng nhau

In [38]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=12)

chọn tham số random_state để tránh thay đổi kết quả sau mỗi lần chạy lại

In [39]:
fig=go.Figure(data=[
                    go.Bar(name=iris.target_names[0], x=['train_set', 'test_set'], y=[(y_train==0).sum(), (y_test==0).sum()]),
                    go.Bar(name=iris.target_names[1], x=['train_set', 'test_set'], y=[(y_train==1).sum(), (y_test==1).sum()]),
                    go.Bar(name=iris.target_names[2], x=['train_set', 'test_set'], y=[(y_train==2).sum(), (y_test==2).sum()])            
])
fig.update_layout(barmode='group')
fig.show()

Tập dữ liệu sau khi chia xảy ra hiện tượng mất cân bằng giữa các class

In [40]:
 iris.data[:,0]

array([5.1, 4.9, 4.7, 4.6, 5. , 5.4, 4.6, 5. , 4.4, 4.9, 5.4, 4.8, 4.8,
       4.3, 5.8, 5.7, 5.4, 5.1, 5.7, 5.1, 5.4, 5.1, 4.6, 5.1, 4.8, 5. ,
       5. , 5.2, 5.2, 4.7, 4.8, 5.4, 5.2, 5.5, 4.9, 5. , 5.5, 4.9, 4.4,
       5.1, 5. , 4.5, 4.4, 5. , 5.1, 4.8, 5.1, 4.6, 5.3, 5. , 7. , 6.4,
       6.9, 5.5, 6.5, 5.7, 6.3, 4.9, 6.6, 5.2, 5. , 5.9, 6. , 6.1, 5.6,
       6.7, 5.6, 5.8, 6.2, 5.6, 5.9, 6.1, 6.3, 6.1, 6.4, 6.6, 6.8, 6.7,
       6. , 5.7, 5.5, 5.5, 5.8, 6. , 5.4, 6. , 6.7, 6.3, 5.6, 5.5, 5.5,
       6.1, 5.8, 5. , 5.6, 5.7, 5.7, 6.2, 5.1, 5.7, 6.3, 5.8, 7.1, 6.3,
       6.5, 7.6, 4.9, 7.3, 6.7, 7.2, 6.5, 6.4, 6.8, 5.7, 5.8, 6.4, 6.5,
       7.7, 7.7, 6. , 6.9, 5.6, 7.7, 6.3, 6.7, 7.2, 6.2, 6.1, 6.4, 7.2,
       7.4, 7.9, 6.4, 6.3, 6.1, 7.7, 6.3, 6.4, 6. , 6.9, 6.7, 6.9, 5.8,
       6.8, 6.7, 6.7, 6.3, 6.5, 6.2, 5.9])

In [41]:
fig = px.scatter( x = iris.data[:,0], y = iris.data[:,1], color=iris.target[:], template="plotly")
fig.show()

Dễ dàng nhìn thấp được sự phân tách tuyến tính của class 0 (Setosa)

#Huấn luyện hệ thống

Sử dụng phương pháp Logistic Regression

In [42]:
from sklearn.linear_model import LogisticRegression
model=LogisticRegression()
model.fit(X_train,y_train)

LogisticRegression()

In [43]:
y_pred=model.predict(X_test)

In [44]:
print(classification_report(y_test,y_pred))

              precision    recall  f1-score   support

           0       1.00      0.92      0.96        12
           1       0.70      0.78      0.74         9
           2       0.78      0.78      0.78         9

    accuracy                           0.83        30
   macro avg       0.83      0.82      0.82        30
weighted avg       0.84      0.83      0.84        30



In [45]:
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score, confusion_matrix, classification_report

print("--Logistic Regression report--")
print("Accuracy: {}".format(accuracy_score(y_test, y_pred)))
print("Precision: ",str(precision_score(y_test, y_pred, average='macro')))
print("Recall: " + str(recall_score(y_test, y_pred, average='macro')))
print("F1_micro: " + str(f1_score(y_test, y_pred, average='micro')))
print("F1_macro: " + str(f1_score(y_test, y_pred, average='macro')))



--Logistic Regression report--
Accuracy: 0.8333333333333334
Precision:  0.825925925925926
Recall: 0.8240740740740741
F1_micro: 0.8333333333333334
F1_macro: 0.8237138740571234


Sử dụng phương pháp KNeighbors Classifier

In [46]:
from sklearn.neighbors import KNeighborsClassifier
range_k=np.arange(1,15)
scores={}
score_list=[]
for k in range_k:
  classifier=KNeighborsClassifier(n_neighbors=k)
  classifier.fit(X_train,y_train)
  y_pred=classifier.predict(X_test)
  scores[k]=f1_score(y_test,y_pred,average='macro')
  score_list.append(scores[k])

In [47]:
score_list

[0.7811421670117323,
 0.7087395696091349,
 0.7077294685990339,
 0.6794643613865583,
 0.6757033248081842,
 0.6759834368530021,
 0.7515891177218408,
 0.7493961352657005,
 0.8518518518518517,
 0.8142414860681114,
 0.8441558441558442,
 0.8142414860681114,
 0.85,
 0.8518518518518517]

In [48]:
np.where(score_list==np.max(score_list))

(array([ 8, 13]),)

Huấn luyện mô hình dựa theo tham số n_neighbors tốt nhất tìm được

In [49]:
classifier=KNeighborsClassifier(n_neighbors=score_list.index(np.max(score_list))+1)
classifier.fit(X_train,y_train)
y_pred=classifier.predict(X_test)

from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score
print("-- KNeighbors Report-")
print("Accuracy: {}".format(accuracy_score(y_test, y_pred)))
print("Precision: ",str(precision_score(y_test, y_pred, average='macro')))
print("Recall: " + str(recall_score(y_test, y_pred, average='macro')))
print("F1_micro: " + str(f1_score(y_test, y_pred, average='micro')))
print("F1_macro: " + str(f1_score(y_test, y_pred, average='macro')))

-- KNeighbors Report-
Accuracy: 0.8666666666666667
Precision:  0.8518518518518517
Recall: 0.8518518518518517
F1_micro: 0.8666666666666667
F1_macro: 0.8518518518518517


Accuracy của mô hình KNeighbors Classifier nhìn chung cao hơn so với Logistic Regression 

Sử dụng phương pháp GridSearchCV để lựa chọn tham số

In [50]:
param_grid=dict(n_neighbors=list(range(1,15)))

In [51]:
param_grid

{'n_neighbors': [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14]}

In [52]:
from sklearn.model_selection import GridSearchCV
from sklearn.pipeline import Pipeline
from sklearn.preprocessing import StandardScaler
from sklearn.metrics import f1_score, make_scorer

#pipeline=Pipeline([('scaler', StandardScaler()),
                   #('model' , KNeighborsClassifier())])
clf=GridSearchCV(KNeighborsClassifier(),param_grid, 
                 cv=7, 
                 scoring=make_scorer(f1_score, average='macro'),
                 verbose=3.0)
clf.fit(X_train,y_train)
print(clf.best_estimator_)
print(clf.best_score_)



Fitting 7 folds for each of 14 candidates, totalling 98 fits
[CV 1/7] END .....................n_neighbors=1;, score=0.720 total time=   0.0s
[CV 2/7] END .....................n_neighbors=1;, score=0.697 total time=   0.0s
[CV 3/7] END .....................n_neighbors=1;, score=0.812 total time=   0.0s
[CV 4/7] END .....................n_neighbors=1;, score=0.667 total time=   0.0s
[CV 5/7] END .....................n_neighbors=1;, score=0.832 total time=   0.0s
[CV 6/7] END .....................n_neighbors=1;, score=0.832 total time=   0.0s
[CV 7/7] END .....................n_neighbors=1;, score=0.657 total time=   0.0s
[CV 1/7] END .....................n_neighbors=2;, score=0.664 total time=   0.0s
[CV 2/7] END .....................n_neighbors=2;, score=0.651 total time=   0.0s
[CV 3/7] END .....................n_neighbors=2;, score=0.878 total time=   0.0s
[CV 4/7] END .....................n_neighbors=2;, score=0.625 total time=   0.0s
[CV 5/7] END .....................n_neighbors=2;

In [53]:
y_clf=clf.predict(X_test)

In [54]:
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score
print("Accuracy: {}".format(accuracy_score(y_test, y_clf)))
print("Precision: ",str(precision_score(y_test, y_clf, average='macro')))
print("Recall: " + str(recall_score(y_test, y_clf, average='macro')))
print("F1_micro: " + str(f1_score(y_test, y_clf, average='micro')))
print("F1_macro: " + str(f1_score(y_test, y_clf, average='macro')))

Accuracy: 0.8666666666666667
Precision:  0.8974358974358975
Recall: 0.8518518518518517
F1_micro: 0.8666666666666667
F1_macro: 0.8441558441558442


-Nhận xét:
  +Mô hình có tham số do GridSearchSV lựa chọn có phần kém hiệu quả hơn so với pháp lựa chọn tham số ở trên. Một phân là do, ở phương pháp thực hiện trên em thực hiện lựa chọn tham số dựa theo mô hình có điểm f1_macro cao nhất đánh giá trên tập test nên tham số sẽ tối ưu nhất cho tập test. 

  +GridSearchSV sẽ thực hiện thử hết các giá trị tham số và Cross Validation trên tập train nên việc lựa chọn tham số cho mô hình có thể được xem là tối ưu nhất 

  +Tuy nhiên, tập dataset trên chỉ có 150 điểm dữ liệu, đây là con số quá nhỏ để có thể huấn luyện mô hình và đánh giá chuẩn xác.

Sử dụng GridSearchCV lựa chọn thông số cho mô hình Logistic Regression

In [55]:
from sklearn.model_selection import GridSearchCV
from sklearn.metrics import f1_score, make_scorer
from sklearn.linear_model import LogisticRegression


param_grid=dict(solver=['saga'],
                penalty=['elasticnet'],
                max_iter=[2000],
                C=list(np.arange(0.2, 2.2, 0.2)),
                l1_ratio=list([0,0.2,0.4,0.6,0.8,1]))

clf=GridSearchCV(LogisticRegression(),param_grid, 
                 cv=5, 
                 scoring=make_scorer(f1_score, average='macro'),
                 verbose=3.0)

clf.fit(X_train,y_train)
print(clf.best_estimator_)
print(clf.best_score_)

Fitting 5 folds for each of 60 candidates, totalling 300 fits
[CV 1/5] END C=0.2, l1_ratio=0, max_iter=2000, penalty=elasticnet, solver=saga;, score=0.833 total time=   0.1s
[CV 2/5] END C=0.2, l1_ratio=0, max_iter=2000, penalty=elasticnet, solver=saga;, score=0.750 total time=   0.1s
[CV 3/5] END C=0.2, l1_ratio=0, max_iter=2000, penalty=elasticnet, solver=saga;, score=0.833 total time=   0.0s
[CV 4/5] END C=0.2, l1_ratio=0, max_iter=2000, penalty=elasticnet, solver=saga;, score=0.764 total time=   0.0s
[CV 5/5] END C=0.2, l1_ratio=0, max_iter=2000, penalty=elasticnet, solver=saga;, score=0.838 total time=   0.0s
[CV 1/5] END C=0.2, l1_ratio=0.2, max_iter=2000, penalty=elasticnet, solver=saga;, score=0.833 total time=   0.0s
[CV 2/5] END C=0.2, l1_ratio=0.2, max_iter=2000, penalty=elasticnet, solver=saga;, score=0.750 total time=   0.0s
[CV 3/5] END C=0.2, l1_ratio=0.2, max_iter=2000, penalty=elasticnet, solver=saga;, score=0.833 total time=   0.0s
[CV 4/5] END C=0.2, l1_ratio=0.2, ma

In [56]:
y_clf=clf.predict(X_test)

In [57]:
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score

print("-- Logistic Regression + GridSearchCV Report --")
print("Accuracy: {}".format(accuracy_score(y_test, y_clf)))
print("Precision: ",str(precision_score(y_test, y_clf, average='macro')))
print("Recall: " + str(recall_score(y_test, y_clf, average='macro')))
print("F1_micro: " + str(f1_score(y_test, y_clf, average='micro')))
print("F1_macro: " + str(f1_score(y_test, y_clf, average='macro')))

-- Logistic Regression + GridSearchCV Report --
Accuracy: 0.9
Precision:  0.9166666666666666
Recall: 0.8888888888888888
F1_micro: 0.9
F1_macro: 0.8857142857142857


Điểm số giữa các than đo Accuracy, Precision, Recall, F1_macro cải thiện nhiều so với sử dụng mô hình Logistic Regression gốc. Thậm chí mô hình này còn tốt hơn cả mô hình KNeigbors Classifier được thử nghiệm ở trên.